File: python/easy/add_binary.py
# Given two binary strings a and b, return their sum as a binary string.

 

# Example 1:

# Input: a = "11", b = "1"
# Output: "100"
# Example 2:

# Input: a = "1010", b = "1011"
# Output: "10101"
 

# Constraints:

# 1 <= a.length, b.length <= 104
# a and b consist only of '0' or '1' characters.
# Each string does not contain leading zeros except for the zero itself.


def addBinary(self, a: str, b: str) -> str:
    return bin(int(a,2)+int(b,2))[2:]


# solution worked on top of the built-in big integer arithmetic and base conversion utilities, it has been taken into account three main principles of the problem:

# 1. both binary strings can be interpreted as integers by explicitly specifying base 2, delegating the positional value computation to the language runtime
# 2. once converted to integers, the binary sum reduces to a single arithmetic addition, leveraging arbitrary-precision integer support to handle carries implicitly
# 3. the resulting integer can be transformed back into its binary string representation, and the prefix added by the conversion utility is removed to match the expected format

# [!] notice that this approach avoids manual bit-by-bit simulation and carry propagation, since all low-level binary operations are abstracted by the language
# [!] correctness relies on the language supporting arbitrary-precision integers; in environments with fixed-width integers, overflow could occur

# this leads to a direct transformation pipeline: binary string -> integer -> integer addition -> binary string


# time complexity: O(n) -> conversion of both strings and output generation = O(n)
# space complexity: O(n) -> storage of intermediate integer representation and result string = O(n)

File: c/easy/add_binary.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

char* addBinary(const char* a, const char* b) {
    int lenA = strlen(a);
    int lenB = strlen(b);
    int maxLen = lenA > lenB ? lenA : lenB;

    char* result = malloc(maxLen + 2);  // +1 for possible carry, +1 for '\0'
    result[maxLen + 1] = '\0';

    int carry = 0;
    int i = lenA - 1, j = lenB - 1, k = maxLen;

    while (k >= 0) {
        int sum = carry;

        if (i >= 0) sum += a[i--] - '0';
        if (j >= 0) sum += b[j--] - '0';

        result[k--] = (sum % 2) + '0';
        carry = sum / 2;
    }

    if (result[0] == '0') {
        memmove(result, result + 1, maxLen + 1);
    }

    return result;
}
